styx-common/src/main/java/com/spotify/styx/util/ShardedCounter.java

```diff
@@ -0,0 +1,241 @@
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Random;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * Implementation of a resource counter on Datastore. Sharded in a way to support increment,
+ * decrement and capping of a counter to a limit, with increased throughput (compared to the 1 write
+ * per second limit on a single entity group) and strong consistency.
+ *
+ * <p>Note that getCounter is supported specifically _without_ strong consistency.
+ *
+ * <p>Note that the ShardedCounter stores state in the same Datastore database that the caller of this
+ * class can access, too; guarantees don't apply if something else than ShardedCounter updates the
+ * stored state.
+ */
+public class ShardedCounter {
+
+  // Ought to be enough (parallelism) for everyone. We could make it dynamic with extra effort.
+  private static final int NUM_SHARDS = 128;
+  private static final long CACHE_EXPIRY_MILLIS = 1000;
+
+  private final Datastore datastore;
+  public static final String KIND_COUNTER_LIMIT = "CounterLimit";
+  public static final String PROPERTY_LIMIT = "limit";
+  public static final String KIND_COUNTER_SHARD = "CounterShard";
+  public static final String PROPERTY_SHARD = "shard";
+
+  private static class CounterSnapshot {
+    Instant updatedAt;
+    Long limit;
+    ArrayList<Long> shards;
+
+    /**
+     * Idempotent initialization, so that we don't reset an existing shard to zero - counterId may
+     * have already been initialized and incremented by another process.
+     */
+    private static void initialize(Datastore datastore, String counterId) {
+      for (int i = 0; i < NUM_SHARDS; i++) {
+        final Key shardKey = datastore.newKeyFactory().setKind(KIND_COUNTER_SHARD)
+            .newKey(counterId + "-" + i);
+        datastore.runInTransaction(transaction -> {
+          final Entity shard = transaction.get(shardKey);
+          if (shard == null) {
+            transaction.put(Entity.newBuilder(shard).set(PROPERTY_SHARD, 0).build());
+          }
+          return null;
+        });
+      }
+    }
+
+    /**
+     * Gets the counter state. Creates it in Datastore if it doesn't exist yet.
+     */
+    private static CounterSnapshot fromDatastore(Datastore datastore, String counterId) {
+      CounterSnapshot snapshot = new CounterSnapshot();
+
+      final Key limitKey = datastore.newKeyFactory().setKind(KIND_COUNTER_LIMIT).newKey(counterId);
+      final Entity limit = datastore.get(limitKey);
+      if (limit == null) {
+        snapshot.limit = Long.MAX_VALUE;
+        // Or IllegalStateException("No limit found in Datastore for " + counterId);?
+      } else {
+        snapshot.limit = limit.getLong(PROPERTY_LIMIT);
+      }
+
+      final EntityQuery queryShards = EntityQuery.newEntityQueryBuilder()
+          .setKind(KIND_COUNTER_SHARD)
+          .setFilter(CompositeFilter.and(
+              PropertyFilter.ge("__key__", counterId + "-0"),
```

Comment:
Could the counter id be a field on the `CounterShard` entity kind? Then this filter could just be `PropertyFilter.eq("counterId", counterId)`.